src/ui/actions/network.ts

```diff
@@ -120,10 +121,21 @@ export function fetchFrames(tsPoint: TimeStampedPoint): UIThunkAction {
   };
 }
 
-export function showRequestDetails(requestId: RequestId) {
-  return {
-    type: "SHOW_REQUEST_DETAILS",
-    requestId,
+export function showRequestDetails(requestId: RequestId): UIThunkAction {
+  return (dispatch, getState) => {
+    const state = getState();
+    const request = state.network.requests.find(request => request.id === requestId);
+    const loadedRegions = getLoadedRegions(state);
+
+    // Don't select a request that's not within a loaded region.
+    if (!request || !loadedRegions || !isPointInRegions(loadedRegions.loaded, request.point)) {
+      return;
+    }
```

Comment:
This filtering logic was previously done in a React component's effect. This seemed bad because (a) it's an extra, unnecessary render and (b) it could cause view inconsistencies and/or flickering if that component happened to load later.

Reply:
I could be wrong, but it seems like if we're going to be pulling in `getState()` anyways, this would make a lot of sense as a check in a reducer. So we would always dispatch the action, but sometimes it would have an effect and other times it would not.

Reply:
> I could be wrong, but it seems like if we're going to be pulling in `getState()` anyways, this would make a lot of sense as a check in a reducer. So we would always dispatch the action, but sometimes it would have an effect and other times it would not.

This is an interesting point to consider, I think– when logic should be run– at the point when the user performs an action or during render?

I think it should be done at the point the user performs an action (aka in the "thunk"). That's the only time when the UI the user has just interacted with is known to match the user's intent. Given that React might delay before rendering, might render more than once, might render a higher priority thing before a lower priority one, etc– filtering in render could mean an unexpected result in some cases.

Waiting to do the filtering in a reducer is akin to waiting until render time (at least for React's `useReducer` and `useState` updater functions). That actually _might not be true for Redux_. I'm not very familiar with it so I don't know.

That's why I moved it to the thunk though. cc @markerikson for thought about this.

Reply:
Oh interesting, my mental model was that redux would process the action and update state as soon as the action was dispatched and then render happens in response to the redux state changing, but it sounds like you're suggesting it could be the other way around in some scenarios? Where react decides to render and that causes the state to update? I'm having trouble picturing how that would work but maybe I'm misunderstanding.

Reply:
Dispatching an action is 100% synchronous by default.  When you call `dispatch(someAction)`, by the time it returns:

- The reducer was run
- The store has saved the new state value internally
- The store has notified all UI subscribers

The mini Redux store shown at https://redux.js.org/tutorials/fundamentals/part-4-store#inside-a-redux-store is a 100% valid running example, and in there the `dispatch` method is just:

```js
  function dispatch(action) {
    state = reducer(state, action)
    listeners.forEach(listener => listener())
  }
```

Next we get into the conceptual question of "how do I handle logic in Redux?"

It's entirely valid to dispatch an action that results in 0 state updates.  We encourage thinking of actions as ["events that happened in an app"](https://redux.js.org/style-guide/#model-actions-as-events-not-setters), which any reducer _may_ respond to if appropriate.  This means that it's possible for _no_ reducers to care about an action.

If that happens, the store subscribers ( `connect` and `useSelector`) will still be called, but they both have eager bailouts that skip doing any further checks if the root state hasn't changed.

On the flip side, it's also totally fine to have logic in a thunk or similar that decides _if_ an action should be dispatched or not based on certain conditions.

Reply:
> but it sounds like you're suggesting it could be the other way around in some scenarios?

No. I don't know what guarantees Redux provides in this case (which is why I tagged Mark). My mental model around reducers is built on React's reducer behavior.

Reply:
thanks both, it's good to know that there is some difference between the redux and react models (because I have always assumed they are super close to each other), and your explanation lines up with what I expect @markerikson, thanks for the thorough explanation! I'm fine with this living in either place. My own preference would be to stick it in the reducer, but my reasons are small and kind of nitpicky, so I'm fine either way.